src/ur5_simulation/ur5_kinematics.py

```diff
@@ -20,55 +21,150 @@ def __init__(self, d1=0.089159, a2=-0.42500, a3=-0.39225, d4=0.1091,
                       [0, 0, 0, 1]])       
  
-    def get_forward_kinematics(self, q=[0, 0, 0, 0, 0, 0], a=0, b=6, short=True):
+    ##############  FORWARD KINEMATICS CALCULATION  ############## 
+
+    def calculate_forward_kinematics(self, q=[0, 0, 0, 0, 0, 0], a=0, b=6, short=True):
         
         '''Receives joint angles vector q and outputs transformation
         matrix from link a to b'''
         
-        self._t1 = q[0]
-        self._t2 = q[1]
-        self._t3 = q[2]
-        self._t4 = q[3]
-        self._t5 = q[4]
-        self._t6 = q[5]
+        self._teta1 = q[0]
+        self._teta2 = q[1]
+        self._teta3 = q[2]
+        self._teta4 = q[3]
+        self._teta5 = q[4]
+        self._teta6 = q[5]
         
         tf = np.eye(4,4)
         for link in range(a,b):
-            tf = np.dot(tf, UR5Kinematics.gen_transf(self.dh_matrix.iloc[link]))
+            tf = np.dot(tf, self._gen_transf(self.dh_matrix.iloc[link]))
         
         if short == True:
             tf = tf.round(3) 
         
         return tf 
-    
\ No newline at end of file
+
+    ##############  INVERSE KINEMATICS CALCULATION  ############## 
+
+    def _calculate_teta_1(self, tf, shoulder='left'):
+        p05 = np.dot(tf, np.array([[0],[0],[-self.__d6],[1]])) # origin 5 in respect to 1
+        p05_x = p05[0][0]
+        p05_y = p05[1][0]
+        p05_xy = (p05_x**2 + p05_y**2)**0.5
+        phi_1 = atan2(p05_y,p05_x)
+        phi_2 = acos(self.__d4/p05_xy)
+        self._teta1 = phi_1 + phi_2 + pi/2
+        
+        if shoulder == 'right':
+            self._teta1 = phi_1 - phi_2 + pi/2
+
+        return self._teta1
+        
+    def _calculate_teta_5(self, tf, wrist='up'):
+        p06_x = tf[0][3]
+        p06_y = tf[1][3]
+        self._teta5 = acos((p06_x*sin(self._teta1) - p06_y*cos(self._teta1) - self.__d4)/self.__d6)
+
+        if wrist == 'down':
+            self._teta5 *= -1 
+
+        return self._teta5
+
+    def _calculate_teta_6(self, tf):
+        t60 = inv(tf) # transformation matrix from 6 to 0
+        y = (-t60[1][0]*sin(self._teta1) + t60[1][1]*cos(self._teta1))/sin(self._teta5)
+        x = (t60[0][0]*sin(self._teta1) - t60[0][1]*cos(self._teta1))/sin(self._teta5)
+        self._teta6 = atan2(y,x)
+
+        return self._teta6   
+
+    def _calculate_teta_3(self, tf, elbow ='up'):
+        t01 = self._gen_transf(self.dh_matrix.iloc[0]) # transformation matrix from 0 to 1
+        t45 = self._gen_transf(self.dh_matrix.iloc[4]) # transformation matrix from 4 to 5
+        t56 = self._gen_transf(self.dh_matrix.iloc[5]) # transformation matrix from 5 to 6
+
+        # t14 = tf @ inv(t01) @ inv(t45) @ inv(t56) error
+        # t14 = tf @ inv(t01) @ inv(np.dot(t45,t56)) error
+        # t14 = inv(t01) @ tf @ inv(t45) @ inv(t56) error
+        # Conclusion: the order of multiplication matters!!
+
+        t14 = inv(t01) @ tf @ inv(np.dot(t45,t56)) # transformation matrix from 1 to 4
```

Comment:
I got a error in this line:
```console
Traceback (most recent call last):
  File "/home/mateus/master_ws/catkin_ws/src/ur5_simulation/scripts/joints_controller_node.py", line 6, in <module>
    from ur5_simulation.ur5_kinematics import UR5Kinematics
  File "/home/mateus/master_ws/catkin_ws/src/ur5_simulation/src/ur5_simulation/ur5_kinematics.py", line 118
    t14 = inv(t01) @ tf @ inv(np.dot(t45,t56)) # transformation matrix from 1 to 4
                   ^
SyntaxError: invalid syntax
```
Like seems because this syntax was included in python 3 and the ROS version which we are using use python 2

Reply:
Got it. The following changes may enable the code to be run on python 2:
Old: `t14 = inv(t01) @ tf @ inv(np.dot(t45,t56))` (for python 3)
New: `t14 = np.dot(np.dot(inv(t01),tf),inv(np.dot(t45,t56)))` (for python 2 and 3)

Reply:
Nice!